src/draft06.rs

```diff
@@ -0,0 +1,757 @@
+pub fn decrypt_response(
+    query: &ObliviousDoHMessagePlaintext,
+    response: &ObliviousDoHMessage,
+    secret: OdohSecret,
+) -> Result<ObliviousDoHMessagePlaintext> {
+    if response.msg_type != ObliviousDoHMessageType::Response {
+        return Err(Error::InvalidMessageType);
+    }
+
+    let response_nonce = response
+        .key_id
+        .as_ref()
+        .try_into()
+        .map_err(|_| Error::InvalidResponseNonceLength)?;
+    let (key, nonce) = derive_secrets(secret, query, response_nonce)?;
+    let cipher = Aes128Gcm::new(GenericArray::from_slice(&key));
+    let mut data = response.encrypted_msg.to_vec();
+
+    let aad = build_aad(ObliviousDoHMessageType::Response, &response.key_id)?;
+
+    cipher
+        .decrypt_in_place(GenericArray::from_slice(&nonce), &aad, &mut data)
+        .map_err(Error::from)?;
+
+    let response_decrypted = parse(&mut data.into())?;
+    Ok(response_decrypted)
+}
+
+/// Decrypt a client query.
+pub fn decrypt_query(
+    query: &ObliviousDoHMessage,
+    key_pair: &ObliviousDoHKeyPair,
+) -> Result<(ObliviousDoHMessagePlaintext, OdohSecret)> {
+    if query.msg_type != ObliviousDoHMessageType::Query {
+        return Err(Error::InvalidMessageType);
+    }
+
+    let key_id = key_pair.public().identifier()?;
+    let key_id_recv = &query.key_id;
+
+    if !key_id_recv.eq(&key_id) {
+        return Err(Error::KeyIdMismatch);
+    }
+
+    let server_sk = key_pair.private();
+    let key_size = <Kex as KeyExchange>::PublicKey::size();
+    let (enc, ct) = query.encrypted_msg.split_at(key_size);
+
+    let encapped_key = EncappedKey::<Kex>::from_bytes(enc).map_err(Error::from)?;
+
+    let mut recv_ctx = hpke::setup_receiver::<Aead, Kdf, Kem>(
+        &OpModeR::Base,
+        &server_sk,
+        &encapped_key,
+        LABEL_QUERY,
+    )
+    .map_err(Error::from)?;
+
+    let aad = build_aad(ObliviousDoHMessageType::Query, &key_id)?;
+
+    let (ciphertext, tag_bytes) = ct.split_at(ct.len() - AEAD_TAG_SIZE);
+    let mut ciphertext_copy = ciphertext.to_vec();
+    let tag = AeadTag::<Aead>::from_bytes(tag_bytes).map_err(Error::from)?;
+
+    recv_ctx
+        .open(&mut ciphertext_copy, &aad, &tag)
+        .map_err(Error::from)?;
+
+    let mut odoh_secret = OdohSecret::default();
+    recv_ctx
+        .export(LABEL_RESPONSE, &mut odoh_secret)
+        .map_err(Error::from)?;
+
+    let plaintext = ciphertext_copy;
+
+    let query_decrypted = parse(&mut plaintext.into())?;
+    Ok((query_decrypted, odoh_secret))
+}
+
+/// Encrypt a server response.
+pub fn encrypt_response(
+    query: &ObliviousDoHMessagePlaintext,
+    response: &ObliviousDoHMessagePlaintext,
+    secret: OdohSecret,
+    response_nonce: ResponseNonce,
+) -> Result<ObliviousDoHMessage> {
+    let buf = compose(response)?;
+
+    let (key, nonce) = derive_secrets(secret, query, response_nonce)?;
+    let cipher = Aes128Gcm::new(GenericArray::from_slice(&key));
+    let mut data = buf.to_vec();
```

Comment:
This will clone the composed response again

Reply:
Unfortunately, the crypto interfaces only accept `Vec<u8>`. The data copy can be avoided here but not everywhere.

Reply:
You can probably implement aead::Buffer for BytesMut, but then ObliviousDoHMessage would have to be defined for owned type (BufMut) for `encrypted_msg` field. This specialization would only work in ObliviousDoHMessage could be deserialized from BytesMut.